File: libs/caller-id-listener/android/src/main/java/expo/modules/calleridlistener/CallSession.kt
package expo.modules.calleridlistener

class CallSession(
    val callId: String,
    val callerNumber: String,
    private val ephemeralToken: String,
    private val fallbackNumber: String?,
    private val onAITranscript: (String, String, String) -> Unit,
    private val onAIToolCall: (String, String, String, String) -> Unit,
    private val onError: (String) -> Unit,
    private val onStatusChange: (String) -> Unit
) {
    var state = "INVITE_RECEIVED"
    
    private var rtpSession: RtpSession? = null
    private var realtimeClient: RealtimeClient? = null
    
    // Parsed from incoming SDP
    var remoteIp: String = ""
    var remotePort: Int = 0
    
    var initialInviteContent: String = ""
    var remoteSipIp: java.net.InetAddress? = null
    var remoteSipPort: Int = 0

    fun parseSdp(content: String) {
        val lines = content.split("\r\n", "\n")
        for (line in lines) {
            if (line.startsWith("c=IN IP4 ")) {
                remoteIp = line.substring("c=IN IP4 ".length).trim()
            } else if (line.startsWith("m=audio ")) {
                val parts = line.substring("m=audio ".length).trim().split(" ")
                if (parts.isNotEmpty()) {
                    remotePort = parts[0].toIntOrNull() ?: 0
                }
            }
        }
    }

    fun startRtpAndRealtime() {
        if (remoteIp.isEmpty() || remotePort == 0) return
        
        realtimeClient = RealtimeClient(ephemeralToken, fallbackNumber, onAudioDelta = { audioBytes ->
            // Audio from OpenAI -> send to phone
            rtpSession?.sendAudio(audioBytes)
        }, onTranscript = onAITranscript, onAIToolCall = onAIToolCall, onError = { msg -> onError("Call $callId: $msg") })
        
        rtpSession = RtpSession(remoteIp, remotePort) { audioBytes ->
            // Audio from phone -> send to OpenAI
            realtimeClient?.sendAudio(audioBytes)
        }
        
        rtpSession?.start()
        realtimeClient?.connect(callId)
        
        state = "ACTIVE"
        onStatusChange("ACTIVE")
    }

    fun getLocalRtpPort(): Int {
        return rtpSession?.localPort ?: 0
    }

    fun sendToolOutput(toolCallId: String, output: String) {
        realtimeClient?.sendToolOutput(toolCallId, output)
    }

    fun stop() {
        state = "TERMINATED"
        rtpSession?.stop()
        realtimeClient?.disconnect()
        onStatusChange("TERMINATED")
    }
}

File: libs/caller-id-listener/ios/CallSession.swift
import Foundation
import Network

class CallSession {
    let callId: String
    let callerNumber: String
    private let ephemeralToken: String
    private let fallbackNumber: String?
    private let onAITranscript: (String, String, String) -> Void
    private let onAIToolCall: (String, String, String, String) -> Void
    private let onError: (String) -> Void
    private let onStatusChange: (String) -> Void
    
    var state = "INVITE_RECEIVED"
    
    private var rtpSession: RtpSession?
    private var realtimeClient: RealtimeClient?
    
    var remoteIp: String = ""
    var remotePort: UInt16 = 0
    
    var initialInviteContent: String = ""
    var sipConnection: NWConnection? = nil
    
    init(callId: String,
         callerNumber: String,
         ephemeralToken: String,
         fallbackNumber: String?,
         onAITranscript: @escaping (String, String, String) -> Void,
         onAIToolCall: @escaping (String, String, String, String) -> Void,
         onError: @escaping (String) -> Void,
         onStatusChange: @escaping (String) -> Void) {
        self.callId = callId
        self.callerNumber = callerNumber
        self.ephemeralToken = ephemeralToken
        self.fallbackNumber = fallbackNumber
        self.onAITranscript = onAITranscript
        self.onAIToolCall = onAIToolCall
        self.onError = onError
        self.onStatusChange = onStatusChange
    }
    
    func parseSdp(_ content: String) {
        let lines = content.components(separatedBy: .newlines)
        for line in lines {
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            if trimmed.hasPrefix("c=IN IP4 ") {
                remoteIp = String(trimmed.dropFirst("c=IN IP4 ".count))
            } else if trimmed.hasPrefix("m=audio ") {
                let parts = trimmed.dropFirst("m=audio ".count).components(separatedBy: " ")
                if let first = parts.first, let port = UInt16(first) {
                    remotePort = port
                }
            }
        }
    }
    
    func startRtpAndRealtime() {
        guard !remoteIp.isEmpty, remotePort != 0 else { return }
        
        realtimeClient = RealtimeClient(
            ephemeralToken: ephemeralToken,
            fallbackNumber: fallbackNumber,
            onAudioDelta: { [weak self] audioBytes in
                self?.rtpSession?.sendAudio(pcmuBytes: audioBytes)
            },
            onTranscript: onAITranscript,
            onAIToolCall: onAIToolCall,
            onError: { [weak self] msg in
                self?.onError("Call \(self?.callId ?? ""): \(msg)")
            }
        )
        
        rtpSession = RtpSession(remoteIp: remoteIp, remotePort: remotePort) { [weak self] audioBytes in
            self?.realtimeClient?.sendAudio(g711Bytes: audioBytes)
        }
        
        rtpSession?.start()
        realtimeClient?.connect(callId: callId)
        
        state = "ACTIVE"
        onStatusChange("ACTIVE")
    }
    
    func getLocalRtpPort() -> UInt16 {
        return rtpSession?.localPort ?? 0
    }
    
    func sendToolOutput(toolCallId: String, output: String) {
        realtimeClient?.sendToolOutput(toolCallId: toolCallId, output: output)
    }
    
    func stop() {
        state = "TERMINATED"
        rtpSession?.stop()
        realtimeClient?.disconnect()
        onStatusChange("TERMINATED")
    }
}
